File: js/deck.js
module.exports = class Deck {

  constructor(shuffle=true) {
    this.reset(shuffle);
  }

  get() {
    return this.cards;
  }
  
  static getIndex(a) {
    return [2, 3, 4, 5, 6, 7, 8, 9, 10, 'J', 'Q', 'K', "A"].indexOf(a)

  }
  /**
   * cut the deck, get a random card.
   * @param {Bool} remove should be removed
   * @returns {Object} card returns a card object
   */
  getRandom(remove=true) {
    const numb = Math.round(Math.random()*(this.cards.length-1));

    const card = this.cards[numb]

    if (remove) this.cards.splice(numb, 1);

    return card;

  }

  /**
   * shuffle the deck, randomize it
   * @returns {Array} list of decks cards 
   */
  shuffle() {
    for (let i = this.cards.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
       
    } 
    return this.cards;
  }
  /**
   * Reset the deck to the original state.
   * @param {Bool} Shuffle on reset 
   * @returns {Deck} the deck object
   * 
   */
  reset(shuffle) {
    const suites = ["️♥️", "♦️", "♠️", "♣️"];
    const cards = [2, 3, 4, 5, 6, 7, 8, 9, 10, 'J', 'Q', 'K', "A"];
    this.cards = [];

    for (const suit of suites) {
      for (const card of cards) {
        this.cards.push({suit: suit, card: card});
      }
    }

    if(shuffle) this.shuffle();

  }
}

File: java/Deck.java
import java.util.*;
public class Deck {
	Random gen = new Random();
	List<HashMap> cards = new ArrayList<HashMap>();

	public Deck(Boolean... shuffle) {
		boolean shuf = shuffle.length > 0 ? shuffle[0] : true;
		
		this.reset(shuf);
		
	}

	void shuffle() {
		Collections.shuffle(this.cards);
	}

	HashMap getRandom(Boolean... remove) {

		boolean rem = remove.length > 0 ? remove[0] : true;

		int cardID = gen.nextInt(this.cards.size());
		HashMap card = this.cards.get(cardID);

		if (rem) this.cards.remove(cardID);

		return card;

	}

	List getCards() {
		return this.cards;
	}

	static int getIndex(String card) {
		String[] cardValues = new String[] {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};

		for (int i=0;i<cardValues.length;i++) {
			if (cardValues[i].equals(card)) {
					return i;
			}
		}
		return -1;
	}


	void reset(Boolean... shuffle) {
		boolean shuf = shuffle.length > 0 ? shuffle[0] : true;

		String[] suites = new String[] {"hearts", "diamonds", "spades", "clubs"};

		String[] cardvalues = new String[] {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};

		for (String suit: suites) {
			for (String card: cardvalues) {
				HashMap<String, String> newCard = new HashMap<>();
				newCard.put("suit", suit);
				newCard.put("card", card);
				this.cards.add(newCard);
			}
		}

		if (shuf) this.shuffle();
	}


}
